lib/vector-core/src/buffers/disk/leveldb_buffer/writer.rs

```diff
@@ -0,0 +1,215 @@
+use super::Key;
+use bytes::Bytes;
+use futures::{task::AtomicWaker, Sink};
+use leveldb::database::{
+    batch::{Batch, Writebatch},
+    options::WriteOptions,
+    Database,
+};
+use std::convert::{TryFrom, TryInto};
+use std::fmt::Debug;
+use std::pin::Pin;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc, Mutex,
+};
+use std::task::{Context, Poll, Waker};
+
+pub struct Writer<T>
+where
+    T: Send + Sync + Unpin + TryInto<Bytes> + TryFrom<Bytes>,
+    <T as TryInto<bytes::Bytes>>::Error: Debug,
+    <T as std::convert::TryFrom<bytes::Bytes>>::Error: Debug,
+{
+    pub(crate) db: Option<Arc<Database<Key>>>,
+    pub(crate) offset: Arc<AtomicUsize>,
+    pub(crate) write_notifier: Arc<AtomicWaker>,
+    pub(crate) blocked_write_tasks: Arc<Mutex<Vec<Waker>>>,
+    pub(crate) writebatch: Writebatch<Key>,
+    pub(crate) batch_size: usize,
+    pub(crate) max_size: usize,
+    pub(crate) current_size: Arc<AtomicUsize>,
+    pub(crate) slot: Option<T>,
+}
+
+// Writebatch isn't Send + Send, but the leveldb docs explicitly say that it's
+// okay to share across threads
+unsafe impl<T> Send for Writer<T>
+where
+    T: Send + Sync + Unpin + TryInto<Bytes> + TryFrom<Bytes>,
+    <T as TryInto<bytes::Bytes>>::Error: Debug,
+    <T as TryFrom<bytes::Bytes>>::Error: Debug,
+{
+}
+unsafe impl<T> Sync for Writer<T>
+where
+    T: Send + Sync + Unpin + TryInto<Bytes> + TryFrom<Bytes>,
+    <T as TryInto<bytes::Bytes>>::Error: Debug,
+    <T as TryFrom<bytes::Bytes>>::Error: Debug,
+{
+}
```

Comment:
How come we need `Sync` now? Was it auto implemented before?

Reply:
It was auto-implemented before. I don't love needing to declare `Sync` directly and I would like to see if there's some tweak to the type logic that'll get us back to auto-implemented. I'll leave this comment open as a TODO.

Reply:
Good call. Removed in 7d2501f